lib/observability/tracing/tracing.go

```diff
@@ -228,9 +242,37 @@ func NewTraceProvider(ctx context.Context, cfg Config) (*Provider, error) {
 			sdktrace.WithResource(res),
 			sdktrace.WithSpanProcessor(sdktrace.NewBatchSpanProcessor(exporter)),
 		),
+		bufferedClient: bufClient,
 	}
 
 	otel.SetTracerProvider(provider)
 
 	return provider, nil
 }
+
+// ReportDelayedResourceAttrs reports resource attributes that were not
+// available when the Provider was created. The context may be used to upload
+// buffered spans to the upstream client.
+func (p *Provider) ReportDelayedResourceAttrs(ctx context.Context, attrs []attribute.KeyValue) error {
+	if p.bufferedClient == nil {
+		return trace.BadParameter("ReportDelayedResourceAttrs called without setting cfg.WaitForDelayedAttributes")
+	}
+	return trace.Wrap(p.bufferedClient.reportDelayedResourceAttrs(ctx, attrs))
+}
+
+// ClosableClient is an [otlptrace.Client] implementation that can also be closed.
+type ClosableClient interface {
+	otlptrace.Client
+	io.Closer
+}
+
+// ReportDelayedClient reports a tracing client to the Provider that was not
+// available when the Provider was created. Ownership is transferred to the
+// Provider, it should not be closed by the caller. The context may be used to
+// upload buffered spans to the client.
+func (p *Provider) ReportDelayedClient(ctx context.Context, clt ClosableClient) error {
```

Comment:
It wasn't super obvious at first to me what `ReportDelayedClient` meant - although I'm struggling to come up with a better name. Perhaps just `SetClient()` ?